### PR TITLE
Issue #235 mitchell 2D

### LINKED ISSUE #235: Mitchels best candidate algorithm should work for 2D cases as well
@Lonnekevb

One of the steps involved in making Flownet "layering" aware is that we need to run the Mitchell's best-candidate algorithm for 2D cases. We need to test whether the current implementation works in these cases, and potentially add it.

See: https://github.com/equinor/flownet/blob/master/src/flownet/network_model/_mitchell.py

To be able to efficiently test this I suggest to write python tests for this code here (both 2D and 3D, neither exist): https://github.com/equinor/flownet/tree/master/tests (for example `test_mitchell.py`).

Additionally, I'd like to ask you to switch out the print statements in the Mitchell's best-candidate algorithm with python logging. Any other improvements in the code are welcomed.

- [ ] Get going
  - [ ] Create a fork of FlowNet
  - [ ] Clone your fork (use SSH), add equinor/flownet as upstream remote
  - [ ] Install FlowNet (see README), install FlowNet.
  - [ ] Clone (your own fork) of the FlowNet-Testdata repo
  - [ ] Try to run the Egg CI model that you can find in the testdata-repo (`flownet ahm ./config/assisted_history_matching.yml ./some_ahm_run --update-config ./ci_config/assisted_history_matching.yml`)
  - [ ] Create a branch on your FlowNet fork and do (some) work
  - [ ] Create a draft PR to show your progress to other team members (don't wait until you're completely done)
- [ ] Implement python tests for the Mitchell's best-candidate algorithm.
- [ ] Change print-statements to python logging. (`import logging`)
- [ ] Improve Mitchell's best-candidate algorithm where needed:
  - [ ] Support 2D cases, if not already ok (maybe we should change the name of the function, to reflect it is not limited to 3D).
  - [ ] Any other problems identified after testing?
  - [ ] General style improvements?
- [ ] Make sure black, pylint and mypym, and (new) tests pass.
- [ ] Change PR status from draft to normal PR. Ask for a code review, iterate potentially.
- [ ] Merge changes after review and CI status is ✔️

### PR DESCRIPTION
**Pull request by LonnekevB**
*Insert a description of your pull request (PR) here, and check off the boxes below when they are done.*

---

### Contributor checklist

- [ ] :tada: This PR closes #235.
- [ ] :scroll: I have broken down my PR into the following tasks:
   - [ ] Task 1
   - [ ] Task 2
- [ ] :robot: I have added tests, or extended existing tests, to cover any new features or bugs fixed in this PR.
- [ ] :book: I have considered adding a new entry in `CHANGELOG.md`.
- [ ] :books: I have considered updating the documentation.

### CODE REVIEW

**Comment by LonnekevB** [comment]
@wouterjdb 
Status update: 

- Tests implemented for 2D and 3D Mitchells best candidate. 

- Adjustments made to the _mitchells.py code to make it work also for 2D. This works, but the 2D issue is still popping up further on in the code (_network_model.py > def _create_connection_groups) where / (maxz - minz) is a division by 0. Haven't had the time to look into this part of the code yet, so haven't fixed this yet.

- Some issues with changing print to logging. Do we want to do this in the whole code? Than I suggest to make that a separate issue and do it in a neat way for the whole code. And do you want to log it to a file or still just to the terminal?

- Having some issues with the CI tests in github. It currently fails on style and linting, but as I don't get this error locally I am having some trouble to figure out what the issue is.

**Comment by wouterjdb** [comment]
> * Adjustments made to the _mitchells.py code to make it work also for 2D. This works, but the 2D issue is still popping up further on in the code (_network_model.py > def _create_connection_groups) where / (maxz - minz) is a division by 0. Haven't had the time to look into this part of the code yet, so haven't fixed this yet.

**Comment by wouterjdb** [comment]
Woops, didn't want to close. Tried to cancel my message only...

> * Tests implemented for 2D and 3D Mitchells best candidate.

☑️ 

> * Adjustments made to the _mitchells.py code to make it work also for 2D. This works, but the 2D issue is still popping up further on in the code (_network_model.py > def _create_connection_groups) where / (maxz - minz) is a division by 0. Haven't had the time to look into this part of the code yet, so haven't fixed this yet.

This will be part of follow-up work. You could write an issue on this. No need to include it in this PR.

> * Some issues with changing print to logging. Do we want to do this in the whole code? Than I suggest to make that a separate issue and do it in a neat way for the whole code. And do you want to log it to a file or still just to the terminal?

Good point. Yes, make it into an issue. Should we not change to logging at all then in this PR?

> * Having some issues with the CI tests in github. It currently fails on style and linting, but as I don't get this error locally I am having some trouble to figure out what the issue is.

If you install a fresh virtual environment on an older version of Ubuntu you might also have an old version of pip. I think that could lead to not getting the latest version of something like black/pylint (bit of guessing here). Try to update pip:

`pip install --upgrade pip`

And update black, pylint and mypy afterwords.

`pip install --upgrade bacl pylint mypy`

 Then run the tests:

```bash
black src tests
pylint src/flownet
mypy --ignore-missing-imports src
```

**Comment by LonnekevB** [comment]
@wouterjdb : The Brugge CI test runs fine locally. But I am currently running it in python 3.6.9. The CI test that fails is the python 3.7 one. I will try running it in python 3.7 tomorrow.

File: src/flownet/network_model/_mitchell.py [hunk 1]
```diff
@@ -47,7 +47,7 @@ def mitchell_best_candidate_modified_3d(
     """
     np.random.seed(random_seed)
     start = time.time()
-    print("Adding flow nodes:  0%", end="")
+    print("Adding flow nodes:  0%", end=" ")
```
Comment (wouterjdb) [inline src/flownet/network_model/_mitchell.py:50]:
```suggestion
    print("Adding flow nodes:  0%", end="")
```

File: src/flownet/network_model/_mitchell.py [hunk 2]
```diff
@@ -88,7 +88,12 @@ def mitchell_best_candidate_modified_3d(
         z_max = max(z)
 
     # Determine the convex hull of the original or linearly scaled perforations
-    hull = Delaunay(np.column_stack([x, y, z]))
+    if np.all([i == z[0] for i in z]):
```
Comment (wouterjdb) [inline src/flownet/network_model/_mitchell.py:91]:
```suggestion
    if np.all(z==z[0]):
```

File: src/flownet/network_model/_mitchell.py [hunk 3]
```diff
@@ -149,7 +154,10 @@ def mitchell_best_candidate_modified_3d(
 
             else:
                 # Test whether all points are inside the convex hull of the perforations
-                in_hull = hull.find_simplex(candidates) >= 0
+                if np.all([i == z[0] for i in z]):
```
Comment (wouterjdb) [inline src/flownet/network_model/_mitchell.py:157]:
```suggestion
                if np.all(z==z[0]):
```

File: src/flownet/network_model/_mitchell.py [hunk 4]
```diff
@@ -163,9 +171,12 @@ def mitchell_best_candidate_modified_3d(
             delta_y_relative = np.power(
                 ((y[0:i] - y_candidate[j]) / (y_max - y_min)), 2
             )
-            delta_z_relative = np.power(
-                ((z[0:i] - z_candidate[j]) / (z_max - z_min)), 2
-            )
+            if np.all([i == z[0] for i in z]):
```
Comment (wouterjdb) [inline src/flownet/network_model/_mitchell.py:174]:
```suggestion
            if np.all(z==z[0]):
```

File: tests/test_mitchell.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
+from typing import List
+
+import numpy as np
+from flownet.network_model._mitchell import mitchell_best_candidate_modified_3d
```
Comment (wouterjdb) [inline tests/test_mitchell.py:4]:
```suggestion

from flownet.network_model._mitchell import mitchell_best_candidate_modified_3d
```

File: tests/test_mitchell.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            random_seed=999,
+        )
+    ]
+
+    x_wells = [x[0] for x in well_perforations_3d]
+    y_wells = [y[1] for y in well_perforations_3d]
+    z_wells = [z[2] for z in well_perforations_3d]
+
+    assert len(coordinates) == (len(well_perforations_3d) + num_added_flow_nodes)
+    assert np.all([x[0] >= min(x_wells) for x in coordinates])
```
Comment (wouterjdb) [inline tests/test_mitchell.py:42]: Same as before, you don't need the list comprehension.

**Review by wouterjdb — CHANGES_REQUESTED** [review]
Looks good. Some minor comments.

NB. I reran the CI yesterday at the end of the day with a fix for Flow 2020.10 and then is was 🟢. The fact that it worked for you locally is that you (and all of us) is that a new version of Flow was released yesterday (happens ~2 times a year, so a bit unlucky there).

File: tests/test_mitchell.py (hunk 6 above)
Comment (LonnekevB) [reply tests/test_mitchell.py:43]: I think here I do, as coordinates is a list of tuples [(x,y,z), (x,y,z)...]. I could move the list comprehension out of the np.all() and first extract the x, y and z coordinates in separate lists. But that just results in extra lines of code...

Comment (wouterjdb) [reply tests/test_mitchell.py:43]: OK, just leave it like this then. 👍

**Review by wouterjdb — APPROVED** [review]
LGTM 🚀

**Pull request merged**